packages/ra-core/src/dataProvider/useDataProvider.ts

```diff
@@ -180,13 +184,18 @@ const useDataProvider = (): DataProviderProxy => {
                         store,
                         type,
                         allArguments,
+                        mutationMode,
                         undoable,
                     };
                     if (isOptimistic) {
                         // in optimistic mode, all fetch calls are stacked, to be
                         // executed once the dataProvider leaves optimistic mode.
                         // In the meantime, the admin uses data from the store.
-                        if (undoable) {
+                        if (
+                            undoable ||
+                            mutationMode === 'undoable' ||
+                            mutationMode === 'optimistic'
+                        ) {
```

Comment:
```suggestion
                        if (
                            undoable ||
                            mutationMode === 'undoable'
                        ) {
```

Reply:
No, I think this change is wrong. We want optimistic calls to be replayed first, just like undoable calls, after existing optimistic mode, and before pessimistic calls. The variable `undoableOptimisticCalls` is badly named, though.